suitcase/spec.py

```diff
@@ -0,0 +1,1064 @@
+    # Not sure how to add the 'detectors' line to the RunStart
+    # md['detectors'] =
+    md['motor_values'] = []
+    md['geometry'] = []
+    line_hash_mapping = {
+        '#G': 'geometry',
+        '#P': 'motor_values',
+    }
+    for line in raw_scan_data[1:]:
+        if line.startswith('#'):
+            # split the line into the "line_type" which tells us the type of
+            # information that this line contains and the "line_contents" which
+            # tells us
+            line_type, line_contents = line.split(' ', 1)
+            if line_type in spec_line_parser:
+                attr, func = spec_line_parser[line_type]
+                md[attr] = func(line_contents)
+            elif line_type == '#L':
+                # It is critical that this line be split on *two* spaces
+                col_names = line_contents.split('  ')
+                md['x_name'] = col_names[0]
+                md['col_names'] = col_names
+            elif line_type[:2] in line_hash_mapping:
+                # These lines are numbered like #G0, #G1, #G2, etc..
+                # We need to just grab the #G part, so take only the first two
+                # elements with line_type[:2]
+                vals = [float(v) for v in line_contents.split()]
+                md[line_hash_mapping[line_type[:2]]].extend(vals)
+    # iterate through the lines again and capture just the scan data
+    scan_data = np.asarray([line.split() for line in raw_scan_data
+                           if not line.startswith('#') if line])
+    try:
+        x = scan_data[:,0]
+    except IndexError:
+        # there must be no scan data...
+        return md, None
+
+    scan_data = pd.DataFrame(
+        data=scan_data, columns=md['col_names'], index=x, dtype=float)
+    scan_data.index.name = md['x_name']
+    return md, scan_data
+
+
+class Specfile:
+    """Object model for a spec file.
+
+    Slicing works on the scan number.
+
+    Attributes
+    ----------
+    filename : str
+        The filename of the spec file that this model represents
+    header : list
+    parsed_header : dict
+        Dictionary of metadata that is contained in the header
+    scans : dict
+        Dictionary of suitcase.spec.Specscan objects where the keys are the
+        scan number
+
+    Usage
+    -----
+    >>> sf = Specfile('/path/to/specfile')
+    >>> scan1 = sf[1]
+    >>> scan1_data = scan1.scan_data
+
+    Notes
+    -----
+    known keys for parsed_header include:
+        - motor_human_names
+        - motor_spec_names
+        - detector_human_names
+        - detector_spec_names
+
+    See the source of `suitcase.spec.parse_spec_scan()` and
+    `suitcase.spec.spec_line_parser` for more complete information
+    """
+    def __init__(self, filename):
+        """
+
+        Parameters
+        ----------
+        filename : str
+            The filename of the spec file that this model represents
+        """
+        self.filename = os.path.abspath(filename)
```

Comment:
IIUC you're suggesting this:

```python
if isinstance(filename, six.string_types):
    filename = os.path.abspath(filename)
self.filename = filename
```

because if filename is a StringIO object then that is also a perfectly valid thing to pass to `open()` but, as the code is currently written will cause an exception to be raised?